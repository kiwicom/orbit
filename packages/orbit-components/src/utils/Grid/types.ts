// Type definitions for @kiwicom/orbit-components
// Project: http://github.com/kiwicom/orbit

import * as React from "react";

import * as Common from "../../common/types";

export interface BasicProps {
  readonly inline?: boolean;
  readonly rows?: string;
  readonly columns?: string;
  readonly gap?: string;
  readonly rowGap?: string;
  readonly columnGap?: string;
  readonly maxWidth?: string;
  readonly width?: string;
}

export interface Props extends Common.Globals, Common.SpaceAfter, BasicProps {
  readonly as?: string;
  readonly mediumMobile?: BasicProps;
  readonly largeMobile?: BasicProps;
  readonly tablet?: BasicProps;
  readonly desktop?: BasicProps;
  readonly largeDesktop?: BasicProps;
  readonly children: React.ReactNode;
}
