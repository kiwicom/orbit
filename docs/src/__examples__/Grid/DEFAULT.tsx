import React from "react";
import { Text, Grid } from "@kiwicom/orbit-components";
import defaultTheme from "@kiwicom/orbit-components/lib/defaultTheme";

export default {
  Example: () => {
    const divStyle = {
      background: `${defaultTheme.orbit.paletteCloudNormal}`,
      border: `1px solid ${defaultTheme.orbit.paletteProductDark}`,
      "box-sizing": "border-box",
      padding: `${defaultTheme.orbit.paddingButtonSmall}`,
    };

    return (
      <Grid columns="repeat(2, 1fr)" rows="repeat(2, 1fr)" rowGap="10px" columnGap="10px">
        <div style={divStyle}>
          <Text>Column 1, Row 1</Text>
        </div>
        <div style={divStyle}>
          <Text>Column 2, Row 1</Text>
        </div>
        <div style={divStyle}>
          <Text>Column 1, Row 2</Text>
        </div>
        <div style={divStyle}>
          <Text>Column 2, Row 2</Text>
        </div>
      </Grid>
    );
  },
  exampleKnobs: [
    {
      component: "Grid",
      knobs: [
        {
          name: "as",
          type: "text",
          defaultValue: "div",
        },
        {
          name: "gap",
          type: "text",
          defaultValue: "",
        },
        {
          name: "rows",
          type: "text",
          defaultValue: "",
        },
        {
          name: "rowGap",
          type: "text",
          defaultValue: "10px",
        },
        {
          name: "width",
          type: "text",
          defaultValue: "",
        },
        {
          name: "inline",
          type: "boolean",
          defaultValue: false,
        },
        {
          name: "columns",
          type: "text",
          defaultValue: "repeat(2, 1fr)",
        },
        {
          name: "columnGap",
          type: "text",
          defaultValue: "10px",
        },
      ],
    },
  ],
};
