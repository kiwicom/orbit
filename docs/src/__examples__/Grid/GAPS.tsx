import React from "react";
import { Grid, Stack, Text, Heading } from "@kiwicom/orbit-components";
import defaultTheme from "@kiwicom/orbit-components/lib/defaultTheme";

export default {
  Example: () => {
    const divStyle: React.CSSProperties = {
      backgroundColor: `${defaultTheme.orbit.paletteCloudDark}`,
      border: `1px solid ${defaultTheme.orbit.paletteProductDark}`,
      boxSizing: "border-box",
      padding: `${defaultTheme.orbit.paddingButtonSmall}`,
    };

    return (
      <Stack>
        <Stack spacing="XSmall">
          <Heading as="h3" type="title3">
            No gaps
          </Heading>
          <Grid columns="repeat(3, 1fr)" rows="repeat(3, 1fr)">
            <div style={divStyle}>
              <Text>Column 1, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 3</Text>
            </div>
          </Grid>
        </Stack>
        <Stack spacing="XSmall">
          <Heading as="h3" type="title3">
            Column gaps
          </Heading>
          <Grid columns="repeat(3, 1fr)" rows="repeat(3, 1fr)" columnGap="32px">
            <div style={divStyle}>
              <Text>Column 1, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 3</Text>
            </div>
          </Grid>
        </Stack>
        <Stack spacing="XSmall">
          <Heading as="h3" type="title3">
            Row gaps
          </Heading>
          <Grid columns="repeat(3, 1fr)" rows="repeat(3, 1fr)" rowGap="32px">
            <div style={divStyle}>
              <Text>Column 1, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 3</Text>
            </div>
          </Grid>
        </Stack>
        <Stack spacing="XSmall">
          <Heading as="h3" type="title3">
            Both in one property
          </Heading>
          <Grid columns="repeat(3, 1fr)" rows="repeat(3, 1fr)" gap="32px">
            <div style={divStyle}>
              <Text>Column 1, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 1</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 2</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 1, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 2, Row 3</Text>
            </div>
            <div style={divStyle}>
              <Text>Column 3, Row 3</Text>
            </div>
          </Grid>
        </Stack>
      </Stack>
    );
  },
  info: {
    title: "Gaps",
    description: "Control the space between your columns and rows with the gaps properties.",
  },
};
