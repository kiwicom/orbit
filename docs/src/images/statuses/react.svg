<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M22.75 11.9075C22.75 10.5826 21.0908 9.32695 18.547 8.54831C19.134 5.95556 18.8731 3.89278 17.7235 3.23237C17.4585 3.07746 17.1487 3.00408 16.8103 3.00408V3.91317C16.9979 3.91317 17.1487 3.94986 17.2751 4.01916C17.8295 4.33714 18.07 5.5479 17.8825 7.10518C17.8376 7.48838 17.7643 7.89197 17.6746 8.30371C16.8756 8.10803 16.0032 7.9572 15.0859 7.85936C14.5356 7.10518 13.9648 6.4203 13.39 5.82104C14.719 4.58581 15.9665 3.90909 16.8144 3.90909V3C15.6933 3 14.2257 3.79902 12.7418 5.18508C11.2579 3.80718 9.79036 3.01631 8.66928 3.01631V3.9254C9.51315 3.9254 10.7647 4.59804 12.0937 5.82511C11.5229 6.42438 10.9522 7.10518 10.41 7.85936C9.48869 7.9572 8.61629 8.10803 7.81727 8.30779C7.7235 7.90012 7.6542 7.50469 7.60528 7.12556C7.41368 5.56828 7.65012 4.35752 8.20047 4.03547C8.32277 3.96209 8.48176 3.92947 8.66928 3.92947V3.02038C8.32684 3.02038 8.01702 3.09376 7.74796 3.24868C6.60243 3.90909 6.3456 5.96779 6.93671 8.55238C4.40104 9.3351 2.75 10.5866 2.75 11.9075C2.75 13.2324 4.40919 14.488 6.95302 15.2666C6.36598 17.8594 6.62689 19.9221 7.7765 20.5826C8.04148 20.7375 8.35131 20.8108 8.69374 20.8108C9.81482 20.8108 11.2824 20.0118 12.7663 18.6258C14.2502 20.0037 15.7178 20.7945 16.8389 20.7945C17.1813 20.7945 17.4911 20.7212 17.7602 20.5662C18.9057 19.9058 19.1626 17.8471 18.5714 15.2625C21.099 14.4839 22.75 13.2283 22.75 11.9075ZM17.4422 9.18834C17.2914 9.71423 17.1039 10.2564 16.8919 10.7986C16.7247 10.4725 16.5494 10.1464 16.3578 9.82022C16.1703 9.49409 15.9705 9.17611 15.7708 8.86629C16.3497 8.9519 16.9082 9.05789 17.4422 9.18834ZM15.5751 13.53C15.2571 14.0803 14.931 14.6021 14.5926 15.0872C13.9852 15.1402 13.3696 15.1688 12.75 15.1688C12.1344 15.1688 11.5189 15.1402 10.9155 15.0913C10.5772 14.6062 10.2469 14.0885 9.92896 13.5422C9.61914 13.0082 9.33785 12.466 9.08102 11.9197C9.33378 11.3734 9.61914 10.8272 9.92489 10.2931C10.2429 9.74276 10.569 9.22095 10.9074 8.73583C11.5148 8.68284 12.1304 8.6543 12.75 8.6543C13.3656 8.6543 13.9811 8.68284 14.5845 8.73176C14.9228 9.21688 15.2531 9.73461 15.571 10.2809C15.8809 10.8149 16.1621 11.3571 16.419 11.9034C16.1621 12.4497 15.8809 12.9959 15.5751 13.53ZM16.8919 13C17.112 13.5463 17.2995 14.0925 17.4544 14.6225C16.9204 14.753 16.3578 14.863 15.7749 14.9486C15.9746 14.6347 16.1744 14.3127 16.3619 13.9825C16.5494 13.6563 16.7247 13.3261 16.8919 13ZM12.7582 17.3498C12.379 16.9584 11.9999 16.5222 11.6248 16.0453C11.9917 16.0616 12.3668 16.0738 12.7459 16.0738C13.1291 16.0738 13.5083 16.0656 13.8792 16.0453C13.5123 16.5222 13.1332 16.9584 12.7582 17.3498ZM9.72513 14.9486C9.14625 14.863 8.58775 14.757 8.05371 14.6266C8.20455 14.1007 8.39207 13.5585 8.60406 13.0163C8.7712 13.3424 8.94649 13.6686 9.1381 13.9947C9.3297 14.3208 9.52538 14.6388 9.72513 14.9486ZM12.7378 6.46514C13.1169 6.8565 13.496 7.2927 13.8711 7.76967C13.5042 7.75336 13.1291 7.74113 12.75 7.74113C12.3668 7.74113 11.9877 7.74929 11.6167 7.76967C11.9836 7.2927 12.3627 6.8565 12.7378 6.46514ZM9.72106 8.86629C9.5213 9.18019 9.32155 9.50224 9.13402 9.83245C8.94649 10.1586 8.7712 10.4847 8.60406 10.8108C8.38392 10.2646 8.19639 9.7183 8.04148 9.18834C8.57552 9.06197 9.1381 8.9519 9.72106 8.86629ZM6.0317 13.9702C4.58857 13.3547 3.65501 12.5475 3.65501 11.9075C3.65501 11.2674 4.58857 10.4562 6.0317 9.84468C6.38229 9.69384 6.76549 9.55932 7.16093 9.43294C7.39329 10.232 7.69904 11.0636 8.07817 11.9156C7.70312 12.7636 7.40145 13.5911 7.17316 14.3861C6.76957 14.2597 6.38636 14.1211 6.0317 13.9702ZM8.22493 19.7958C7.67051 19.4778 7.42998 18.267 7.61751 16.7097C7.66235 16.3265 7.73573 15.923 7.82542 15.5112C8.62444 15.7069 9.49684 15.8577 10.4141 15.9556C10.9644 16.7097 11.5352 17.3946 12.11 17.9939C10.781 19.2291 9.53353 19.9058 8.68559 19.9058C8.50214 19.9018 8.34723 19.8651 8.22493 19.7958ZM17.8947 16.6894C18.0863 18.2466 17.8499 19.4574 17.2995 19.7795C17.1772 19.8528 17.0182 19.8854 16.8307 19.8854C15.9869 19.8854 14.7353 19.2128 13.4063 17.9857C13.9771 17.3865 14.5478 16.7057 15.09 15.9515C16.0113 15.8537 16.8837 15.7028 17.6827 15.5031C17.7765 15.9148 17.8499 16.3102 17.8947 16.6894ZM19.4642 13.9702C19.1136 14.1211 18.7304 14.2556 18.335 14.382C18.1026 13.583 17.7969 12.7513 17.4178 11.8993C17.7928 11.0514 18.0945 10.2238 18.3228 9.42886C18.7264 9.55524 19.1096 9.69384 19.4683 9.84468C20.9114 10.4603 21.845 11.2674 21.845 11.9075C21.8409 12.5475 20.9074 13.3587 19.4642 13.9702Z" fill="#697D95"/>
<path d="M12.7458 13.771C13.7748 13.771 14.6089 12.9369 14.6089 11.908C14.6089 10.879 13.7748 10.0449 12.7458 10.0449C11.7169 10.0449 10.8828 10.879 10.8828 11.908C10.8828 12.9369 11.7169 13.771 12.7458 13.771Z" fill="#697D95"/>
</svg>
