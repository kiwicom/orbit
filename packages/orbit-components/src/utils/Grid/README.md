# Grid

To implement the Grid component in your project, add the import:

```jsx
import Grid from "@kiwicom/orbit-components/lib/utils/Grid";
```

Then you can use it:

```jsx
<Grid columns="4fr minmax(256px, 1fr)">
  // You can use your custom components
  <Section>Hello world!</Section>
  <SideBar>I'm content on the right side.</SideBar>
</Grid>
```

## Props

The table below contains all of the props available in the Grid component.

| Name         | Type                       | Default | Description                                                                                      |
| :----------- | :------------------------- | :------ | :----------------------------------------------------------------------------------------------- |
| **children** | `React.Node`               |         | The content of the Grid. [See Usage](#usage)                                                     |
| className    | `string`                   |         | If needed, you can extend the Grid component with `styled` and add your own styles.              |
| columns      | `string`                   | `"1fr"` | Property alias for `grid-template-columns`.                                                      |
| columnGap    | `string`                   |         | Gap size for `columns`.                                                                          |
| dataTest     | `string`                   |         | Optional prop for testing purposes.                                                              |
| desktop      | [`Object`](#media-queries) |         | Object for properties for desktop viewports. [See Media queries](#media-queries)                 |
| as           | `string`                   | `"div"` | The valid DOM element used for the root node.                                                    |
| gap          | `string`                   |         | Gap size for `columns` and `rows`.                                                               |
| inline       | `boolean`                  | `false` | If `true`, the Grid will have `display: inline-grid`; otherwise it will be `display: grid`.      |
| tablet       | [`Object`](#media-queries) |         | Object for setting up properties for tablet viewports. [See Media queries](#media-queries)       |
| largeDesktop | [`Object`](#media-queries) |         | Object for setting up properties for largeDesktop viewports. [See Media queries](#media-queries) |
| largeMobile  | [`Object`](#media-queries) |         | Object for setting up properties for largeMobile viewports. [See Media queries](#media-queries)  |
| maxWidth     | `string`                   |         | Alias for the `max-width` property of the Grid component.                                        |
| mediumMobile | [`Object`](#media-queries) |         | Object for setting up properties for mediumMobile viewports. [See Media queries](#media-queries) |
| rows         | `string`                   | `"1fr"` | Property alias for `grid-template-rows`.                                                         |
| rowGap       | `string`                   |         | The gap size for `rows`.                                                                         |
| width        | `string`                   |         | Alias for the `width` property of the Grid component.                                            |
| spaceAfter   | `enum`                     |         | Additional `padding` to bottom of the Stack.                                                     |

### Media Queries

When you need to specify specific behaviours for the Grid component on different viewports, use the appropriate properties.
The component has `mediumMobile`, `largeMobile`, `tablet`, `desktop`, and `largeDesktop`, which behave the same as [mediaQuery](https://github.com/kiwicom/orbit/tree/master/packages/orbit-components/src/utils/mediaQuery) functions.
All of the object for these properties have the same properties and none is required.

| Name       | Type      | Default  | Description                                                                      |
| :--------- | :-------- | :------- | :------------------------------------------------------------------------------- |
| columns    | `string`  | `"none"` | Property alias for `grid-template-columns`.                                      |
| columnGap  | `string`  |          | Gap size for `columns`.                                                          |
| gap        | `string`  |          | Gap size for `columns` and `rows`.                                               |
| inline     | `boolean` | `false`  | If `true`, the grid will have `display: inline-grid`; otherwise `display: grid`. |
| maxWidth   | `string`  |          | Alias for the `max-width` property of the Grid component.                        |
| rows       | `string`  | `"none"` | Property alias for `grid-template-rows`.                                         |
| rowGap     | `string`  |          | Gap size for `rows`.                                                             |
| width      | `string`  |          | Alias for the `width` property of the Grid component.                            |
| spaceAfter | `enum`    |          | Additional `padding` to bottom of the Stack.                                     |

### enum

| spaceAfter   |
| :----------- |
| `"none"`     |
| `"smallest"` |
| `"small"`    |
| `"normal"`   |
| `"medium"`   |
| `"large"`    |
| `"largest"`  |

## Why you should use this component

The main focus of this component is to provide an API that solves the most common issues and gives you a quick how to start using CSS grids in your project without worries.

By default, this Grid component handles:

- Compiling the `repeat` function to the old specification
- Auto-placement of its children

## Usage

This component works the same way as a native CSS grid, so it expects the same hierarchy when it comes to using children.

For instance, when you want to define a Grid that has two columns and two rows, the complete usage should look like this:

```jsx
import Grid from "@kiwicom/orbit-components/lib/utils/Grid";

// your custom component to fill each cell in the grid
import Cell from "./Cell";

const Example = () => (
  <Grid columns="1fr 1fr" rows="1fr 1fr">
    <Cell>First column, first row</Cell>
    <Cell>Second column, first row</Cell>
    <Cell>First column, second row</Cell>
    <Cell>Second column, second row</Cell>
  </Grid>
);
```
