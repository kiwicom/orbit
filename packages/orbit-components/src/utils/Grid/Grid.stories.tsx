import * as React from "react";
import styled from "styled-components";
import { text, boolean, number, object, select } from "@storybook/addon-knobs";

import SPACINGS_AFTER from "../../common/getSpacingToken/consts";

import Grid from ".";

const CustomDiv = styled.div`
  background: rgba(0, 169, 145, 0.2);
`;

export default {
  title: "Grid",
};

export const Playground = () => {
  const inline = boolean("inline", false);
  const maxWidth = text("maxWidth", "1440px");
  const width = text("width", "100%");
  const columns = text("columns", undefined);
  const rows = text("rows", "repeat(8, 40px)");
  const gap = text("gap", null);
  const columnGap = text("columnGap", null);
  const rowGap = text("rowGap", "10px");
  const as = text("as", "div");
  const dataTest = text("dataTest", "test");
  const divsCount = number("divsCount", 8);
  const mediumMobile = object("mediumMobile", {
    rowGap: "0",
  });
  const largeMobile = object("largeMobile", {
    columns: "repeat(4, 1fr)",
    rows: "repeat(2, 40px)",
    gap: "20px",
  });
  const tablet = object("tablet", {
    columnGap: "40px",
  });
  const desktop = object("desktop", {
    columns: "repeat(2, minmax(100px, 1fr))",
    rows: "repeat(4, 40px)",
    gap: "40px",
  });
  const largeDesktop = object("largeDesktop", {
    columns: "repeat(8, minmax(10px, 1fr))",
    rows: "40px",
  });
  const spaceAfter = select("spaceAfter", Object.values(SPACINGS_AFTER), SPACINGS_AFTER.NORMAL);
  return (
    <Grid
      inline={inline}
      spaceAfter={spaceAfter}
      maxWidth={maxWidth}
      width={width}
      columns={columns}
      rows={rows}
      gap={gap}
      columnGap={columnGap}
      rowGap={rowGap}
      as={as}
      dataTest={dataTest}
      mediumMobile={mediumMobile}
      largeMobile={largeMobile}
      tablet={tablet}
      desktop={desktop}
      largeDesktop={largeDesktop}
    >
      {Array(...Array(divsCount)).map((_, key) => (
        // eslint-disable-next-line
        <CustomDiv key={key} />
      ))}
    </Grid>
  );
};
