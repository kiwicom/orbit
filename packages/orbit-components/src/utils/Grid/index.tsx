import * as React from "react";
import styled, { css } from "styled-components";

import defaultTheme from "../../defaultTheme";
import mediaQueries from "../mediaQuery";
import { DEVICES } from "../mediaQuery/consts";
import getViewportGridStyles from "./helpers/getViewportGridStyles";
import { isDefined } from "../layout";
import { Props } from "./index.d";

const StyledGrid = styled.div<Props>`
  ${({ theme, ...props }) =>
    Object.values(DEVICES).map(viewport =>
      viewport in mediaQueries
        ? mediaQueries[viewport](css`
            ${isDefined(props[viewport]) &&
            getViewportGridStyles({
              viewport,
              theme,
            })};
          `)
        : viewport === "smallMobile" &&
          css`
            ${getViewportGridStyles({ viewport, theme })};
          `,
    )};
`;

StyledGrid.defaultProps = {
  theme: defaultTheme,
};

const Grid = ({
  inline,
  rows = "1fr",
  columns = "1fr",
  gap,
  rowGap,
  columnGap,
  maxWidth,
  width,
  children,
  dataTest,
  as = "div",
  ...props
}: Props) => {
  const smallMobile = { inline, rows, columns, gap, rowGap, columnGap, maxWidth, width };
  return (
    // @ts-expect-error As prop
    <StyledGrid {...props} smallMobile={smallMobile} data-test={dataTest} as={as}>
      {children}
    </StyledGrid>
  );
};

export default Grid;
