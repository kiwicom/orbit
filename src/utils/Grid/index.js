// @flow
import * as React from "react";
import styled, { css } from "styled-components";

import defaultTheme from "../../defaultTheme";
import mediaQueries from "../mediaQuery";
import { DEVICES } from "../mediaQuery/consts";
import getViewportGridStyles from "./helpers/getViewportGridStyles";
import isDefined from "../../Stack/helpers/isDefined";

import type { Props } from ".";

const StyledGrid = styled(({ className, children, dataTest }) => (
  <div className={className} data-test={dataTest}>
    {children}
  </div>
))`
  /*
    Just apply all mediaQueries (from devices map)
    smallMobile - default values are not mediaQuery and needs to be rendered differently
   */
  ${props =>
    DEVICES.map((viewport, index, devices) =>
      viewport in mediaQueries
        ? mediaQueries[viewport](css`
            ${isDefined(props[viewport]) && getViewportGridStyles({ viewport, index, devices })};
          `)
        : viewport === "smallMobile" &&
          css`
            ${getViewportGridStyles({ viewport, index, devices })};
          `,
    )};
`;

StyledGrid.defaultProps = {
  theme: defaultTheme,
};

const Grid = ({
  inline,
  rows = "1fr",
  columns = "1fr",
  gap,
  rowGap,
  columnGap,
  maxWidth,
  children,
  dataTest,
  element = "div",
  ...props
}: Props) => {
  const smallMobile = { inline, rows, columns, gap, rowGap, columnGap, maxWidth };
  return (
    <StyledGrid smallMobile={smallMobile} data-test={dataTest} as={element} {...props}>
      {children}
    </StyledGrid>
  );
};

export default Grid;
