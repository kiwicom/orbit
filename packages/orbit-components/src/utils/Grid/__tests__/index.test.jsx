// @flow
import * as React from "react";
import { render } from "@testing-library/react";

import Grid from "..";

describe("Grid", () => {
  it("should match snapshot", () => {
    const dataTest = "test";
    const maxWidth = "1440px";
    const width = "100%";
    const columns = "repeat(2, 1fr)";
    const columnGap = "10px";
    const rows = "repeat(2, minmax(30px, 100px))";
    const rowGap = "20px";

    const { container } = render(
      <Grid
        as="div"
        dataTest={dataTest}
        inline
        maxWidth={maxWidth}
        width={width}
        columns={columns}
        columnGap={columnGap}
        rows={rows}
        rowGap={rowGap}
        mediumMobile={{ maxWidth: "600px" }}
        largeMobile={{ maxWidth: "700px" }}
        tablet={{ maxWidth: "800px" }}
        desktop={{ maxWidth: "900px" }}
        largeDesktop={{ maxWidth: "1000px" }}
      >
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </Grid>,
    );

    expect(container).toMatchSnapshot();
  });
});
