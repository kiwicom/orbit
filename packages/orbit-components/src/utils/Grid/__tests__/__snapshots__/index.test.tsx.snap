// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should match snapshot 1`] = `
<div>
  <div
    class="inline-grid w-[var(--grid-width)] max-w-[var(--grid-max-width)] mm:max-w-[var(--mm-grid-max-width)] lm:max-w-[var(--lm-grid-max-width)] tb:max-w-[var(--tb-grid-max-width)] de:max-w-[var(--de-grid-max-width)] ld:max-w-[var(--ld-grid-max-width)] grid-cols-[var(--grid-columns)] grid-rows-[var(--grid-rows)] gap-x-[var(--grid-column-gap)] gap-y-[var(--grid-row-gap)]"
    data-test="test"
    style="--grid-width: 100%; --grid-max-width: 1440px; --mm-grid-max-width: 600px; --lm-grid-max-width: 700px; --tb-grid-max-width: 800px; --de-grid-max-width: 900px; --ld-grid-max-width: 1000px; --grid-columns: repeat(2, 1fr); --grid-rows: repeat(2, minmax(30px, 100px)); --grid-column-gap: 10px; --grid-row-gap: 20px;"
  >
    <div>
      1
    </div>
    <div>
      2
    </div>
    <div>
      3
    </div>
    <div>
      4
    </div>
  </div>
</div>
`;
